### PR TITLE
Add FilesUtil.share_folder method

### PR DESCRIPTION
**Pull request by genestack-solomatin**
Depends on https://github.com/genestack/genestack/pull/3839

### CODE REVIEW

File: genestack_client/files_util.py [hunk 1]
```diff
@@ -376,6 +377,36 @@ def share_files(self, accessions, group, destination_folder=None, password=None)
...
+        :param password: password for sharing,
+               if not specified, will be asked for in an interactive prompt (if possible)
+        :type: str
+        :rtype: None
+        """
+        self.share_files([folder_accession], group, destination_folder=destination_folder, password=password)
+        SudoUtils(self.connection).ensure_sudo_interactive(password)
+        share_utils = self.connection.application('genestack/shareutils')
+        limit = 100
+        delay = 1  # delay between attempts
```
Comment (alexander-polovtcev) [inline genestack_client/files_util.py]: specify the time unit in the comment

File: genestack_client/files_util.py [hunk 2]
```diff
@@ -376,6 +377,36 @@ def share_files(self, accessions, group, destination_folder=None, password=None)
...
+        while True:
+            offset = 0
+            count = share_utils.invoke('shareChunkInFolder', folder_accession, group, offset, limit)
+            if count == 0 and offset == 0:
+                return
+            if count < limit:
+                sleep(delay)
+                offset = 0
+            offset += limit
```
Comment (alexander-polovtcev) [inline genestack_client/files_util.py]: you will never query the first page if `count < limit` condition is true

File: genestack_client/files_util.py [hunk 3]
```diff
@@ -376,6 +377,36 @@ def share_files(self, accessions, group, destination_folder=None, password=None)
...
+        :type: str
+        :rtype: None
+        """
+        self.share_files([folder_accession], group, destination_folder=destination_folder, password=password)
+        SudoUtils(self.connection).ensure_sudo_interactive(password)
+        share_utils = self.connection.application('genestack/shareutils')
+        limit = 100
+        delay = 1  # delay between attempts
+        while True:
+            offset = 0
```
Comment (alexander-polovtcev) [inline genestack_client/files_util.py]: and here you will always query the first page

**Comment by genestack-solomatin** [comment]
This branch should be merged after https://github.com/genestack/genestack/pull/4085

File: genestack_client/files_util.py [hunk 4]
```diff
@@ -380,6 +381,39 @@ def share_files(self, accessions, group, destination_folder=None, password=None)
...
+        :type: str
+        :rtype: None
+        """
+        self.share_files([folder_accession], group, destination_folder=destination_folder, password=password)
+        share_utils = self.connection.application('genestack/shareutils')
+        limit = 100
+        delay_seconds = 1  # delay between attempts
+
+        offset = 0
+        while True:
```
Comment (alexander-polovtcev) [inline genestack_client/files_util.py]: Can you add some comments here about the sharing procedure too? It's quite tedious to look it up in the `shareChunkInFolder` documentation

Comment (genestack-solomatin) [reply genestack_client/files_util.py]: Done

File: genestack_client/files_util.py [hunk 5]
```diff
@@ -380,6 +381,48 @@ def share_files(self, accessions, group, destination_folder=None, password=None)
         if destination_folder is not None:
             share_utils.invoke('linkFiles', accessions, destination_folder, group)
 
+    def share_folder(self, folder_accession, group, destination_folder=None, password=None):
+        """
+        Recursively share folder.
+
+        This method make repeated calls to server, each method call
```
Comment (alexander-polovtcev) [inline genestack_client/files_util.py]: `makes`

File: genestack_client/files_util.py [hunk 6]
```diff
@@ -380,6 +381,48 @@ def share_files(self, accessions, group, destination_folder=None, password=None)
...
+        shares chunk of the files in the folder that aren't shared yet with ``group` or ``WORLD``.
```
Comment (alexander-polovtcev) [inline genestack_client/files_util.py]: `chunk of files`

Comment (alexander-polovtcev) [reply genestack_client/files_util.py]: `in the given folder`

File: genestack_client/files_util.py [hunk 7]
```diff
@@ -380,6 +381,48 @@ def share_files(self, accessions, group, destination_folder=None, password=None)
...
+
+        Due to the indexing lag, same files can be included in different calls.
+        Method implementation is written to overcome this limitation.
+        Sum of the counts returned by all call is greater of equal than the count of files shared.
```
Comment (alexander-polovtcev) [inline genestack_client/files_util.py]: I don't understand this sentence

File: genestack_client/files_util.py [hunk 8]
```diff
@@ -380,6 +381,48 @@ def share_files(self, accessions, group, destination_folder=None, password=None)
...
+
+        Method can work for a long time if ``folder`` contains a lot of files.
```
Comment (alexander-polovtcev) [inline genestack_client/files_util.py]: what's this for?

Comment (genestack-solomatin) [reply genestack_client/files_util.py]: It can work for couple of hours without any output.

**Pull request merged**